lib/Auth.js

```diff
@@ -66,21 +66,45 @@ function useAuthProvider() {
 
         await fetchUserProfileAndStartClient();
 
-        // Loop through all existing authentication providers and sign in
-        Object.entries(activeAuthentications).forEach(([type, authentication]) => {
-            // Skip the default Matrix auth provider
-            if (type === 'matrix') return;
-            authentication.signin(username, password).catch(() => { });
-        });
+        await getAccessTokensFromAccountData(username, password);
 
         return matrixAuthenticationResponse;
     };
 
+    const getAccessTokensFromAccountData = async (username, password) => {
+        // we will use `getAccountDataFromServer` instead of `getAccountData` as we can not tell for sure if the sync is completely done and dont want to wait for that. So potentially this call is avoidable but it is more manageable to handle it like this.
+        const accessTokens = await getAuthenticationProvider('matrix').getMatrixClient().getAccountDataFromServer('dev.medienhaus.spaces.accesstokens') || {};
+        // fetch did not worked or this accountData state event is not set yet. In either way we want to create it.
+        let dataChanged = false;
+
+        for await (const [type, authentication] of Object.entries(activeAuthentications)) {
+            // Skip the default Matrix auth provider
+            if (type === 'matrix') continue;
+
+            if (accessTokens[type]) {
+                authentication.addToken(accessTokens[type]);
+                const tokenValidation = await authentication.validateToken();
+                if (tokenValidation) {
+                    // everything seems to work out as expected we will cut here
+                    continue;
+                }
+            } else {
+                // if the stored token does not work anymore then we will execute a login procedure and store the new generated accesstoken
+                dataChanged = true;
+                accessTokens[type] = await authentication.signin(username, password).catch(() => { });
+            }
+        }
+
+        if (dataChanged) {
+            // write accessTokensToAccountData. this function is needed to prevent spamming with eventId's, even if the accesstokens are all still valid and doesnt need to be updated.
```

Comment:
To be clear: that last part of the comment

> this function is needed to prevent spamming with eventId's, even if the accesstokens are all still valid and doesnt need to be updated

refers to the `if (dataChanged)` condition, right? As in, you're trying to say that we only write new access tokens to the Matrix account data *if* and *only if* at least one of the access tokens has changed, right?

If I understood that right, can you clarify that in the documentation? So "function" should probably be "condition", and event IDs do not really play a role here, or do they? Is account data based on events?

Reply:
changed